File: cdc/src/share/basis/classes/common/sun/awt/image/Image.java
/*
 * @(#)Image.java	1.58 06/10/10
 *
 * Copyright  1990-2008 Sun Microsystems, Inc. All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License version
 * 2 only, as published by the Free Software Foundation. 
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details (a copy is
 * included at /legal/license.txt). 
 * 
 * You should have received a copy of the GNU General Public License
 * version 2 along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA 
 * 
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa
 * Clara, CA 95054 or visit www.sun.com if you need additional
 * information or have any questions. 
 *
 */

package sun.awt.image;

import java.util.Hashtable;
import java.util.Enumeration;
import java.awt.Component;
import java.awt.Color;
import java.awt.Graphics;
import java.awt.image.ImageProducer;
import java.awt.image.ImageConsumer;
import java.awt.image.ImageObserver;

public abstract class Image extends java.awt.Image {
    /**
     * The object which is used to reconstruct the original image data
     * as needed.
     */
    ImageProducer source;
    InputStreamImageSource src;
    ImageRepresentation imagerep;
    protected Image() {}

    /**
     * Construct an image for offscreen rendering to be used with a
     * given Component.
     */
    protected Image(Component c, int w, int h) {
        if (w <= 0 || h <= 0)
            throw new IllegalArgumentException ("width and height must be > 0");
        OffScreenImageSource osis = new OffScreenImageSource(c, w, h);
        source = osis;
        width = w;
        height = h;
        properties = new Hashtable();
        availinfo |= (ImageObserver.WIDTH |
                    ImageObserver.HEIGHT |
                    ImageObserver.PROPERTIES);
        imagerep = makeImageRep();
        imagerep.offscreen = true;
        imagerep.setDimensions(w, h);
        if (c != null)
            imagerep.offscreenInit(c.getBackground());
        osis.setImageRep(imagerep);
    }

    /**
     * Construct an image from an ImageProducer object.
     */
    protected Image(ImageProducer is) {
        source = is;
        if (is instanceof InputStreamImageSource) {
            src = (InputStreamImageSource) is;
        }
    }

    public ImageProducer getSource() {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        return source;
    }

    protected void initGraphics(Graphics g) {
        OffScreenImageSource osis = (OffScreenImageSource) source;
        g.setColor(osis.target.getForeground());
        g.setFont(osis.target.getFont());
    }

    public Color getBackground() {
        OffScreenImageSource osis = (OffScreenImageSource) source;
        return osis.target.getBackground();
    }
    private int width = -1;
    private int height = -1;
    private Hashtable properties;
    private int availinfo;
    /**
     * Return the width of the original image source.
     * If the width isn't known, then the image is reconstructed.
     */
    public int getWidth() {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if ((availinfo & ImageObserver.WIDTH) == 0) {
            reconstruct(ImageObserver.WIDTH);
        }
        return width;
    }

    /**
     * Return the width of the original image source.
     * If the width isn't known, then the ImageObserver object will be
     * notified when the data is available.
     */
    public synchronized int getWidth(ImageObserver iw) {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if ((availinfo & ImageObserver.WIDTH) == 0) {
            addWatcher(iw, true);
            if ((availinfo & ImageObserver.WIDTH) == 0) {
                return -1;
            }
        }
        return width;
    }

    /**
     * Return the height of the original image source.
     * If the height isn't known, then the image is reconstructed.
     */
    public int getHeight() {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if ((availinfo & ImageObserver.HEIGHT) == 0) {
            reconstruct(ImageObserver.HEIGHT);
        }
        return height;
    }

    /**
     * Return the height of the original image source.
     * If the height isn't known, then the ImageObserver object will be
     * notified when the data is available.
     */
    public synchronized int getHeight(ImageObserver iw) {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if ((availinfo & ImageObserver.HEIGHT) == 0) {
            addWatcher(iw, true);
            if ((availinfo & ImageObserver.HEIGHT) == 0) {
                return -1;
            }
        }
        return height;
    }

    /**
     * Return a property of the image by name.  Individual property names
     * are defined by the various image formats.  If a property is not
     * defined for a particular image, then this method will return the
     * UndefinedProperty object.  If the properties for this image are
     * not yet known, then this method will return null and the ImageObserver
     * object will be notified later.  The property name "comment" should
     * be used to store an optional comment which can be presented to
     * the user as a description of the image, its source, or its author.
     */
    public Object getProperty(String name, ImageObserver observer) {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if (properties == null) {
            addWatcher(observer, true);
            if (properties == null) {
                return null;
            }
        }
        Object o = properties.get(name);
        if (o == null) {
            o = Image.UndefinedProperty;
        }
        return o;
    }

    public boolean hasError() {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        return (availinfo & ImageObserver.ERROR) != 0;
    }

    public int check(ImageObserver iw) {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if ((availinfo & ImageObserver.ERROR) == 0 &&
            ((~availinfo) & (ImageObserver.WIDTH |
                    ImageObserver.HEIGHT |
                    ImageObserver.PROPERTIES)) != 0) {
            addWatcher(iw, false);
        }
        return availinfo;
    }

    public void preload(ImageObserver iw) {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if ((availinfo & ImageObserver.ALLBITS) == 0) {
            addWatcher(iw, true);
        }
    }

    private synchronized void addWatcher(ImageObserver iw, boolean load) {
        if ((availinfo & ImageObserver.ERROR) != 0) {
            if (iw != null) {
                iw.imageUpdate(this, ImageObserver.ERROR | ImageObserver.ABORT,
                    -1, -1, -1, -1);
            }
            return;
        }
        ImageRepresentation ir = getImageRep();
        ir.addWatcher(iw);
        if (load) {
            ir.startProduction();
        }
    }

    private synchronized void reconstruct(int flags) {
        if ((flags & ~availinfo) != 0) {
            if ((availinfo & ImageObserver.ERROR) != 0) {
                return;
            }
            ImageRepresentation ir = getImageRep();
            ir.startProduction();
            while ((flags & ~availinfo) != 0) {
                try {
                    wait();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
                if ((availinfo & ImageObserver.ERROR) != 0) {
                    return;
                }
            }
        }
    }

    synchronized void addInfo(int newinfo) {
        availinfo |= newinfo;
        notifyAll();
    }

    void setDimensions(int w, int h) {
        width = w;
        height = h;
        addInfo(ImageObserver.WIDTH | ImageObserver.HEIGHT);
    }

    void setProperties(Hashtable props) {
        if (props == null) {
            props = new Hashtable();
        }
        properties = props;
        addInfo(ImageObserver.PROPERTIES);
    }

    synchronized void infoDone(int status) {
        if (status == ImageConsumer.IMAGEERROR ||
            ((~availinfo) & (ImageObserver.WIDTH |
                    ImageObserver.HEIGHT)) != 0) {
            addInfo(ImageObserver.ERROR);
        } else if ((availinfo & ImageObserver.PROPERTIES) == 0) {
            setProperties(null);
        }
    }

    public void flush() {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if (!(source instanceof OffScreenImageSource)) {
            ImageRepresentation ir;
            synchronized (this) {
                availinfo &= ~ImageObserver.ERROR;
                ir = imagerep;
                imagerep = null;
            }
            if (ir != null) {
                ir.abort();
            }
            if (src != null) {
                src.flush();
            }
        }
    }

    protected abstract ImageRepresentation makeImageRep();

    protected synchronized ImageRepresentation getImageRep() {
        if (src != null) {
            src.checkSecurity(null, false);
        }
        if (imagerep == null) {
            imagerep = makeImageRep();
        }
        return imagerep;
    }
}

File: javacall/implementation/win32_emul/midp/image.c
/*
 * Copyright  1990-2008 Sun Microsystems, Inc. All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License version
 * 2 only, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details (a copy is
 * included at /legal/license.txt).
 *
 * You should have received a copy of the GNU General Public License
 * version 2 along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa
 * Clara, CA 95054 or visit www.sun.com if you need additional
 * information or have any questions.
 */


 /**
  * Win32 implementation supports the following image formats: jpeg,png
 */

#include <stdio.h>
#include <windows.h>
#include "javacall_image.h"
#include <jpegdecoder.h>

    /**
     * PNG Header Data
     */
    unsigned char pngHeader[] ={
         (unsigned char)0x89, (unsigned char)0x50, (unsigned char)0x4e, (unsigned char)0x47,
         (unsigned char)0x0d, (unsigned char)0x0a, (unsigned char)0x1a, (unsigned char)0x0a,
         (unsigned char)0x00, (unsigned char)0x00, (unsigned char)0x00, (unsigned char)0x0d,
         (unsigned char)0x49, (unsigned char)0x48, (unsigned char)0x44, (unsigned char)0x52
    };
    int pngHeaderSize = 16;
    /**
     * JPEG Header Data
     */
    unsigned char jpegHeader[] ={
         (unsigned char)0xff, (unsigned char)0xd8, (unsigned char)0xff, (unsigned char)0xe0
    };
    int jpegHeaderSize = 4;

    /**
     * RAW Header Data
     */
    unsigned char rawHeader[] = {
         (unsigned char)0x89, (unsigned char)0x53, (unsigned char)0x55, (unsigned char)0x4E
    };

    int rawHeaderSize = 4;

    typedef struct _imageData{
        int width;
        int height;
        char* pixelData;
        int pixelDataSize;
    }imageData;

/**
 * Function to compare byte data to the given header
 */
javacall_result headerMatch(unsigned char* header,int headerSize,
                                unsigned char* imageData,
                                int imageLength) {
    int i=0;
    if (imageLength < headerSize) {
        return JAVACALL_FAIL;
    }
    for (i = 0; i < headerSize; i++) {
        if (imageData[i] != header[i]) {
            return JAVACALL_FAIL;
        }
    }
    return JAVACALL_OK;
}
javacall_result initPNGImageInfo(unsigned char* source,long sourceSize,javacall_image_info* imageInfo){
    int width = ((source[16] & 0x0ff) << 24) +
                    ((source[17] & 0x0ff) << 16) +
                    ((source[18] & 0x0ff) <<  8) +
                    (source[19] & 0x0ff);

    int height = ((source[20] & 0x0ff) << 24) +
                     ((source[21] & 0x0ff) << 16) +
                     ((source[22] & 0x0ff) <<  8) +
                     (source[23] & 0x0ff);

        if (width <= 0 || height <= 0) {
            return JAVACALL_FAIL;
        }
        imageInfo->width = width;
        imageInfo->height = height;
        return JAVACALL_OK;

}


javacall_result initJPEGImageInfo(unsigned char* source,long sourceSize,javacall_image_info* imageInfo){
        int width = 0;
        int height = 0;
        int field_len = 0;
        int idx = 2;
        unsigned char code;

        while (idx + 8 <sourceSize) {
            if (source[idx] != (unsigned char)0xff) {
                break;
            }

            if ((unsigned char) (source[idx + 1] & 0xf0) == (unsigned char) 0xc0) {
                code = source[idx + 1];

                if (code != (unsigned char) 0xc4 || code != (unsigned char) 0xcc) {
                    /* Big endian */
                    height = ((source[idx + 5] & 0xff) << 8) +
                            (source[idx + 6] & 0xff);
                    width = ((source[idx + 7] & 0xff) << 8) +
                            (source[idx + 8] & 0xff);
                    break;
                }
            }

            /* Go to the next marker */
            field_len = ((source[idx + 2] & 0xff) << 8) +
                    (source[idx + 3] & 0xff);
            idx += field_len + 2;
        }

        if (width <= 0 || height <= 0) {
            return JAVACALL_FAIL;
        }
        imageInfo->width = width;
        imageInfo->height = height;
        return JAVACALL_OK;

}

/**
 * Get image information
 *
 * @param source        Pointer to image data source
 * @param sourceSize    Size of source in bytes
 * @param imageInfo     Pointer to javacall_image_info type buffer
 *
 * @retval JAVACALL_OK
 *          imageData is acceptable format and image information returned by imageInfo argument
 * @retval JAVACALL_FAIL
 *          imageData is not acceptable
 */
javacall_result javacall_image_get_info(const void* source,
                                        long sourceSize,
                                        /*OUT*/ javacall_image_info* imageInfo){
    javacall_result res= JAVACALL_FAIL;
    //check what is the image format
    if (headerMatch(pngHeader,pngHeaderSize,(unsigned char*)source,sourceSize)==JAVACALL_OK){
        res = initPNGImageInfo((unsigned char*)source,sourceSize,imageInfo);
        res = JAVACALL_FAIL;// currently supporting only jpeg via native image decoder

    }else if(headerMatch(jpegHeader,jpegHeaderSize,(unsigned char*)source,sourceSize)==JAVACALL_OK){//JPEG
        res = initJPEGImageInfo((unsigned char*)source,sourceSize,imageInfo);

    }

    return res;
}

/**
 * Start a decoding transaction of the source image to array of javacall_pixel type
 *
 * For Synchronous decoding:
 * 1.Copy the address of the source buffer to a static variable (also the handle can be used for that).
 * 2.Return JAVACALL_OK
 *
 * For Asynchronous decoding:
 * 1.Allocate a memory block of size sourceSize
 * 2.Copy the source buffer to the new allocated buffer.
 * 3.Send the buffer to the decoder which is in another task.
 * 4.Mark the decoding operation with a unique handle.
 * 5.Return JAVACALL_WOULD_BLOCK

 * @param source        Pointer to image data source
 * @param handle        Handle of this image decoding. This could be used from async decoding mode
 *                      It could return a NULL value, if this API return JAVACALL_OK
 *
 *
 * @retval JAVACALL_OK          Decoding OK! - Synchronous decoding
 * @retval JAVACALL_FAIL        Fail
 * @retval JAVACALL_WOULD_BLOCK Image decoding performed asynchronously
 */
javacall_result javacall_image_decode_start(const void* source,
                                            long sourceSize,
                                            long width,
                                            long height,
                                            /*OUT*/ javacall_handle* handle){
    imageData* image= malloc(sizeof(imageData));
    image->width = width;
    image->height = height;
    image->pixelDataSize = sourceSize;
    image->pixelData =(char*) malloc(sourceSize*sizeof(char));
    if (image->pixelData != NULL) {
        memcpy(image->pixelData,source,sourceSize*sizeof(char));
        *handle = (void*)image;
        return JAVACALL_OK;
    }
    return JAVACALL_FAIL;
}

/**
 * Finalize the image decoding transaction
 *
 * For Synchronous decoding:
 *   1.This function will be called right after javacall_image_decode_start() returns with JAVACALL_OK.
 *   2.Decode the source buffer that was stored previously into decodeBuf , alphaBuf.
 * For Asynchronous decoding:
 *   1.This function will be called after javanotify_on_image_decode_end is invoked.
 *   2.The platform should only copy the decoded image to decodeBuf , alphaBuf.
 *
 * @param handle        The handle get from javacall_image_decode_start
 * @param decodeBuf     Pointer to decoding target buffer
 * @param decodeBufSize Size of decodeBuf in bytes
 * @param alphaBuf      Pointer to alpha data buffer
 *                      It could be a NULL if there is no need to get alpha data
 * @param alphaBufSize  Size of alphaBuf in bytes
 *
 * @retval JAVACALL_OK      Success
 * @retval JAVACALL_FAIL    Fail
 */
javacall_result javacall_image_decode_finish(javacall_handle handle,
                                             /*OUT*/ javacall_pixel* decodeBuf,
                                             long decodeBufSize,
                                             /*OUT*/ char* alphaBuf,
                                             long alphaBufSize){

    javacall_result result = JAVACALL_FAIL;
    void *info ;
    imageData* image = (imageData*)handle;

    int i;
    for (i=0;i<alphaBufSize;i++) {//initialize alpha buffer
        alphaBuf[i]=(char)0xFF;
    }
    if (headerMatch(jpegHeader,jpegHeaderSize,image->pixelData,image->pixelDataSize)== JAVACALL_OK) {
        info = JPEG_To_RGB_init();
        if (info) {
            int width, height;
            if (JPEG_To_RGB_decodeHeader(info,image->pixelData, image->pixelDataSize,
                &width, &height) != 0) {
                if (JPEG_To_RGB_decodeData2(info,(char*)decodeBuf,
                    sizeof(unsigned short), 0, 0, image->width, image->height) != 0) {
                    result = JAVACALL_OK;
                }
            }
            JPEG_To_RGB_free(info);
        }
    }else if (headerMatch(pngHeader,pngHeaderSize,image->pixelData,image->pixelDataSize)==JAVACALL_OK) {
         result = JAVACALL_FAIL;
    }
    return result;
}
